sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala

```diff
@@ -58,6 +67,15 @@ case class BoundReference(ordinal: Int, dataType: DataType, nullable: Boolean)
     }
   }
 
+  def name: String = _name.getOrElse(s"i[$ordinal]")
+
+  private var _name: Option[String] = None
```

Comment:
Lets not add mutable state inside of expressions.  There are a lot of places where we cache computation and you are going to make it hard to reason about correctness .